### PR TITLE
Add warnings for duplicate test IDs & misconfigured retry commands

### PR DESCRIPTION
**Pull request by pierrebeaucamp**
* If Captain encounters a test result that contains a duplicate test, it emits a warning.
* If Captain retries a test, but the updated test result doesn't contain the test anymore, it emits a warning.

Both warnings can also be configured as hard failures: Either as CLI Flags (`--fail-on-duplicate-test-id` and `--fail-on-misconfigured-retry`) or in the config file (as `fail-on-duplicate-test-id: true` in the suite config and `fail-on-misconfiguration: true` in the retry config respectively). 

---

Duplicate test identification has an interesting edge-case: Although we can trivially check for duplicates using the `test.Matches` method, we sometimes match tests using an "identity recipe" instead - these recipes usually only use a subset of available test metadata, which leads to the possibility of `test.Matches` not identifying a duplicate while an "identity recipe" would.

To work around this, we'll check for duplicates using the identity-recipe as well - however this requires that we know expected identity recipes ahead of time. To make this work, we can now pull these identity recipes from Cloud. Captain will do this if it cannot find a `.captain/recipes.json` file. 

**Note**: This PR doesn't add any kind of invalidation for the cached recipes. So far, our recipes never had to change. Adding invalidation is not much work should we need to do this in the future.

--- 

I ended up refactoring a small part of the retry code: I noticed that we explicitly checked for `test.Attempt.Status.ImpliesFailure()` in each `SubstitudeFor` method for each framework. Instead, I've put this check inside the `filter` function that we pass along.

There are only two places that call `SubstitudeFor`, so adding `test.Attempt.Status.ImpliesFailure()` into the filter centralized the logic to these two places (instead of duplicating it across each framework). Furthermore, this allowed me to re-use the `filter` function to later check for misconfigured retry commands.

The only adverse effect of this change is that I had to update a bunch of test-files as they had a stub filter function previously.

### CODE REVIEW

File: internal/parsing/parse.go [hunk 1]
```diff
@@ -161,6 +182,63 @@ func parseWith(file fs.File, parsers []Parser, groupNumber int, log *zap.Sugared
...
+			}
+
+			// TODO: What about "other" frameworks (i.e. when we don't match here)
+			if recipe, ok := cfg.IdentityRecipes[framework.String()]; ok {
+				identityRecipe = &recipe
+			}
+
+			break FRAMEWORKS
+		}
+	}
```
Comment (kylekthompson) [inline internal/parsing/parse.go:203]: I think this all is potentially unnecessary - the `finalResults` we grabbed `.Tests` from has `.Framework` on it. Correct me if I'm missing something though 😅

File: internal/parsing/parse.go [hunk 2]
```diff
@@ -161,6 +182,63 @@ func parseWith(file fs.File, parsers []Parser, groupNumber int, log *zap.Sugared
...
+	uniqueTestIdentifiers := make(map[string]struct{})
+
+FRAMEWORKS:
+	for framework, frameworkParsers := range cfg.FrameworkParsers {
+		for _, frameworkParser := range frameworkParsers {
+			if reflect.TypeOf(frameworkParser) != reflect.TypeOf(parser) {
+				continue
+			}
+
+			// TODO: What about "other" frameworks (i.e. when we don't match here)
```
Comment (kylekthompson) [inline internal/parsing/parse.go:196]: We have a fallback in the identity recipes in Cloud - maybe we should maintain that same behavior?

Comment (pierrebeaucamp) [reply internal/parsing/parse.go:196]: Whoops, I wanted to address that still...

File: internal/parsing/parse.go (hunk 1 above)
Comment (pierrebeaucamp) [reply internal/parsing/parse.go:203]: Excellent observation, thank you

File: internal/cli/run.go [hunk 3]
```diff
@@ -503,6 +509,10 @@ func (s Service) attemptRetries(
 		}
 
 		filter := func(test v1.Test) bool {
+			if !test.Attempt.Status.ImpliesFailure() {
+				return false
+			}
```
Comment (kylekthompson) [inline internal/cli/run.go:514]: Can you talk me through this change?

File: internal/cli/run.go [hunk 4]
```diff
@@ -635,6 +645,32 @@ func (s Service) attemptRetries(
 				s.Log.Warn(err)
 			}
 		}
+
+	FLATTENED_TEST_RESULTS:
+		for _, originalTest := range flattenedTestResults.Tests {
```
Comment (kylekthompson) [inline internal/cli/run.go:650]: I wonder if this would be more clear written as `originalTestResults` - I recognize they're the same thing at this point though (as an aside, maybe we should pull both `flattenedTestResults` and `flattenedNewlyExecutedTestResults` down after this new loop where we actually assign values to them and change them

Don't feel like you have to change anything, though - it just took me a sec without the context of the code higher up above this loop

File: internal/parsing/parse.go [hunk 5]
```diff
@@ -134,7 +138,23 @@ func parseWith(file fs.File, parsers []Parser, groupNumber int, log *zap.Sugared
...
+	testIDsAreUnique, err := checkIfTestIDsAreUnique(finalResults, cfg)
+	if err != nil {
+		return nil, err
+	}
+	if !testIDsAreUnique {
+		duplicateTestIDWarningMsg := fmt.Sprintf(
+			"Test result file %q contains two or more tests that share the same metadata (such as name or identifier). %s",
+			file.Name(),
+			"Please make sure test identifiers are unique.",
+		)
```
Comment (kylekthompson) [inline internal/parsing/parse.go:152]: Is there any chance we could get a list of the tests that have issues in here? Can totally be a follow up

File: internal/parsing/parse.go [hunk 6]
```diff
@@ -161,6 +181,46 @@ func parseWith(file fs.File, parsers []Parser, groupNumber int, log *zap.Sugared
 	return &finalResults, nil
 }
 
+func checkIfTestIDsAreUnique(testResult v1.TestResults, cfg Config) (bool, error) {
+	uniqueTestIdentifiers := make(map[string]struct{})
+	identityRecipe, recipeFound := cfg.IdentityRecipes[testResult.Framework.String()]
```
Comment (kylekthompson) [inline internal/parsing/parse.go:186]: Did you decide to add support for the fallback recipe? 

In cloud we use `[file, description], strict?: false`

Comment (pierrebeaucamp) [reply internal/parsing/parse.go:186]: Yes and no - I just noticed a mistake I made while typing up an explanation. Anyway, for context:

Cloud returns a recipe for "other" which is the `[file, description], strict?: false` fallback you mentioned. As long as a framework in Captain can be identified as "other", it will use this fallback.

There is a separate discussion to be had whether or not we always want to use this fallback in case we cannot match a specific recipe. For example, this would happen if Captain has a framework that is not present in Cloud. Personally, I like erroring on the side of caution and just have it no-op in this scenario.

As for the error / mistake a mentioned: I just noticed that `testResult.Framework` will never match against "other" since it's set by one of the available parsers. Only after we return from `parseWith`, we add the coerced Framework (which might be "other") to the test results. I'll fix this.

File: internal/cli/run.go (hunk 4 above)
Comment (pierrebeaucamp) [reply internal/cli/run.go:650]: Hm, I think we need this to be `flattenedTestResults` though.

For the first retry, `originalTestResults` and `flattenedTestResults` are the same. But for subsequent retries, we want to compare results against the merged result set, not the original one.

For example, you might have the following scenario:

**Original**
Test 1: Failure
Test 2: Failure

**Retry 1**
Test 1: Success
Test 2: Failure

**Retry 2**
Test 2: Success

I.e. we do want to merge the original results and the first retry before comparing against the results of the second retry - otherwise Captain could identify "Test 1" to be missing from the second retry, even though it succeeded previously.

**Review by dan-manges — APPROVED** [review]
